samza-core/src/main/java/org/apache/samza/classloader/IsolatingClassLoaderFactory.java

```diff
@@ -0,0 +1,316 @@
+    String baseDirectoryPath = System.getProperty("user.dir");
+    File apiLibDirectory = libDirectory(new File(baseDirectoryPath, DependencyIsolationUtils.FRAMEWORK_API_DIRECTORY));
+    LOG.info("Using API lib directory: {}", apiLibDirectory);
+    File infrastructureLibDirectory =
+        libDirectory(new File(baseDirectoryPath, DependencyIsolationUtils.FRAMEWORK_INFRASTRUCTURE_DIRECTORY));
+    LOG.info("Using infrastructure lib directory: {}", infrastructureLibDirectory);
+    File applicationLibDirectory =
+        libDirectory(new File(baseDirectoryPath, DependencyIsolationUtils.APPLICATION_DIRECTORY));
+    LOG.info("Using application lib directory: {}", applicationLibDirectory);
+
+    ClassLoader apiClassLoader = buildApiClassLoader(apiLibDirectory);
+    ClassLoader applicationClassLoader =
+        buildApplicationClassLoader(applicationLibDirectory, apiLibDirectory, apiClassLoader);
+
+    // the classloader to return is the one with the infrastructure classpath
+    return buildInfrastructureClassLoader(infrastructureLibDirectory, apiLibDirectory, apiClassLoader,
+        applicationClassLoader);
+  }
+
+  /**
+   * Build the {@link ClassLoader} which can load framework API classes.
+   *
+   * This sets up the link between the bootstrap classloader and the API classloader (see {@link #buildClassLoader()}.
+   */
+  private static ClassLoader buildApiClassLoader(File apiLibDirectory) {
+    /*
+     * This can just use the built-in classloading, which checks the parent classloader first and then checks its own
+     * classpath. A null parent means bootstrap classloader, which contains core Java classes (e.g. java.lang.String).
+     * This doesn't need to be isolated from the parent, because we only want to load all bootstrap classes from the
+     * bootstrap classloader.
+     */
+    return new URLClassLoader(getClasspathAsURLs(apiLibDirectory), null);
+  }
+
+  /**
+   * Build the {@link ClassLoader} which can load application classes.
+   *
+   * This sets up the link between the application classloader and the API classloader (see {@link #buildClassLoader()}.
+   */
+  private static ClassLoader buildApplicationClassLoader(File applicationLibDirectory, File apiLibDirectory,
+      ClassLoader apiClassLoader) {
+    return LoaderBuilder.anIsolatingLoader()
+        // look in application lib directory for JARs
+        .withClasspath(getClasspathAsURIs(applicationLibDirectory))
+        // getClasspathAsURIs should only return JARs within applicationLibDirectory anyways, but doing it to be safe
+        .withOriginRestriction(OriginRestriction.denyByDefault().allowingDirectory(applicationLibDirectory, false))
+        // delegate to the api classloader for API classes
+        .withParentRelationship(buildApiParentRelationship(apiLibDirectory, apiClassLoader))
+        .build();
+  }
+
+  /**
+   * Build the {@link ClassLoader} which can load Samza framework core classes.
```

Comment:
To clarify documentation, this can/will load application classes as well, right?